db_stress_tool/expected_state.cc

```diff
@@ -3,39 +3,106 @@
+  const bool during_read_no_write_happened =
+      (pre_read_expected_value_base ==
+       GetValueBase(post_read_expected_value_base + expected_value_base_delta));
+
+  return pre_read_expected_deleted && during_read_no_write_happened;
+}
+
+bool ExpectedState::NotDeletedDuringRead(uint32_t pre_read_expected_value,
+                                         uint32_t post_read_expected_value) {
+  const bool pre_read_expected_not_deleted =
+      !IsDeleted(pre_read_expected_value);
+
+  const uint32_t pre_read_expected_del_counter =
+      GetDelCounter(pre_read_expected_value);
+  const uint32_t post_read_expected_del_counter =
+      GetDelCounter(post_read_expected_value);
+  const uint32_t expected_del_counter_delta =
+      PendingDelete(post_read_expected_value) ? DEL_COUNTER_DELTA : 0;
+  const bool during_read_no_delete_happened =
+      (pre_read_expected_del_counter ==
+       GetDelCounter(post_read_expected_del_counter +
+                     expected_del_counter_delta));
+
+  return pre_read_expected_not_deleted && during_read_no_delete_happened;
+}
+
+bool ExpectedState::InExpectedValueBaseRange(
+    uint32_t value_base, uint32_t pre_read_expected_value_base,
+    uint32_t post_read_expected_value_base,
+    bool post_read_expected_pending_write) {
+  assert(IsValuePartValid(value_base, VALUE_BASE_MASK) &&
+         IsValuePartValid(pre_read_expected_value_base, VALUE_BASE_MASK) &&
+         IsValuePartValid(post_read_expected_value_base, VALUE_BASE_MASK));
+
+  const uint32_t expected_value_base_delta =
+      post_read_expected_pending_write ? VALUE_BASE_DELTA : 0;
+
+  uint32_t post_read_expected_value_base_final =
+      GetValueBase(post_read_expected_value_base + expected_value_base_delta);
+
+  if (pre_read_expected_value_base <= post_read_expected_value_base_final) {
+    const uint32_t lower_bound = pre_read_expected_value_base;
+    const uint32_t upper_bound = post_read_expected_value_base_final;
+    return lower_bound <= value_base && value_base <= upper_bound;
+  } else {
+    const uint32_t upper_bound_1 = post_read_expected_value_base_final;
+    const uint32_t lower_bound_2 = pre_read_expected_value_base;
+    const uint32_t upper_bound_2 = VALUE_BASE_MASK;
+    return (value_base <= upper_bound_1) ||
+           (lower_bound_2 <= value_base && value_base <= upper_bound_2);
+  }
+}
+
 ExpectedState::ExpectedState(size_t max_key, size_t num_column_families)
     : max_key_(max_key),
       num_column_families_(num_column_families),
       values_(nullptr) {}
 
 void ExpectedState::ClearColumnFamily(int cf) {
   std::fill(&Value(cf, 0 /* key */), &Value(cf + 1, 0 /* key */),
-            SharedState::DELETION_SENTINEL);
+            ExpectedState::DEL_MASK);
 }
 
-void ExpectedState::Put(int cf, int64_t key, uint32_t value_base,
-                        bool pending) {
-  if (!pending) {
-    // prevent expected-value update from reordering before Write
-    std::atomic_thread_fence(std::memory_order_release);
```

Comment:
[Question] Why was this needed previously when the ExpectedState::Put holds lock?

Reply:
The comment says to prevent reordering within the thread holding the lock. I forgot (like usual) to write the context of why that matters. As you seem to have already realized, it doesn't matter when other threads must grab the lock before reading or writing the same value. But in the case where crash happens after one of the updates but before the other, the order of those updates matters. In particular, if we allowed reordering to happen such that the expected-value was written before the write operation, crash could come between the two. Then recovery would expect a value to be there that was never written.

Reply:
Fixed by adding std::atomic_thread_fence(std::memory_order_release); back as it was